### PR TITLE
Add BC layers

### PR DESCRIPTION
**Pull request by Ener-Getick**
This PR adds several BC layers.
It also includes the ``FileParam`` backward compatible support.

Classes marked as deprecated:
- ``FOS\RestBundle\Controller\Annotations\Param``
- ``FOS\RestBundle\Request\AbstractRequestBodyParamConverter``
- ``FOS\RestBundle\Request\RequestBodyParamConverter20``
- ``FOS\RestBundle\Util\Codes``
- ``FOS\RestBundle\Util\FormatNegotiator``
- ``FOS\RestBundle\Util\FormatNegotiatorInterface``
- ``FOS\RestBundle\Util\MediaTypeNegotiatorInterface``
- ``FOS\RestBundle\Util\ViolationFormatter``
- ``FOS\RestBundle\Util\ViolationFormatterInterface``

At the same time I'm wondering if we should not refactor the bundle annotations to stop using public properties and instead using getters/setters...

### CODE REVIEW

**Comment by lsmith77** [comment]
not sure what the benefit would be of adding accessors to the annotations

**Comment by lsmith77** [comment]
if we do this .. then maybe it makes more sense after all to do a 1.8 release

File: Controller/Annotations/AbstractParam.php [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+use Symfony\Component\DependencyInjection\ContainerAware;
+
+/**
+ * {@inheritdoc}
+ *
+ * @author Jordi Boggiano <j.boggiano@seld.be>
+ * @author Boris Guéry <guery.b@gmail.com>
+ * @author Ener-Getick <egetick@gmail.com>
+ */
+abstract class AbstractParam extends ContainerAware implements ParamInterface
```
Comment (stof) [inline Controller/Annotations/AbstractParam.php]: a container-aware annotation ? WTF ?

File: Controller/Annotations/FileParam.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+/**
+ * Represents a file that must be present.
+ *
+ * @Annotation
+ * @Target("METHOD")
+ *
+ * @author Ener-Getick <egetick@gmail.com>
+ */
+class FileParam extends AbstractParam
```
Comment (stof) [inline Controller/Annotations/FileParam.php]: This is not a BC layer, it is a new feature. So it must not be in this PR

File: Util/Codes.php [hunk 3]
```diff
@@ -22,6 +22,8 @@
  *
  * @author Jordi Boggiano <j.boggiano@seld.be>
  * @author Markus Lanthaler <markus.lanthaler@gmx.net>
+ *
+ * @deprecated since 1.7, to be removed in 2.0. Use {@link \Symfony\Component\HttpFoundation\Response} instead.
```
Comment (stof) [inline Util/Codes.php]: you should add a ``@trigger_error`` on top

**Comment by Ener-Getick** [comment]
We can't use ``trigger_error`` or throw an exception when someone gets the value directly without using accessors ...
Moreover we can't really check the type of the value.

Maybe can someone give use a hint ?

File: Util/ViolationFormatter.php [hunk 4]
```diff
@@ -11,40 +11,11 @@
...
 
-use FOS\RestBundle\Controller\Annotations\Param;
-use FOS\RestBundle\Controller\Annotations\QueryParam;
-use Symfony\Component\Validator\ConstraintViolationInterface;
-use Symfony\Component\Validator\ConstraintViolationListInterface;
+use FOS\RestBundle\Validator\ViolationFormatter as BaseViolationFormatter;
 
-class ViolationFormatter implements ViolationFormatterInterface
+/**
+ * @deprecated since 1.7, to be removed in 2.0. Use {@link BaseViolationFormatter} instead.
```
Comment (stof) [inline Util/ViolationFormatter.php]: using the FQCN would make the message more understandable when being logged by the DebugClassLoader

**Comment by stof** [comment]
@Ener-Getick usign public properties is the way the annotation parser works

File: Controller/Annotations/AbstractParam.php (hunk 1 above)
Comment (Ener-Getick) [inline Controller/Annotations/AbstractParam.php]: The ``Container`` is used to resolve the parameter in the default value. We should maybe move this to the ``ParamFetcher`` ...

**Comment by Ener-Getick** [comment]
@stof not necessary, for example the ``Route`` annotation uses private properties (see https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Routing/Annotation/Route.php).

File: Controller/Annotations/AbstractParam.php (hunk 1 above)
Comment (Ener-Getick) [inline Controller/Annotations/AbstractParam.php]: I remember why it was not in the ``ParamFetcher`` ... The ``requirements`` are also resolved but this cannot be done in the ``ParamFetcher``.
So we have two solutions:
- remove the ``requirements`` resolution
- let the annotation extending ``ContainerAware``

Comment (stof) [inline Controller/Annotations/AbstractParam.php]: injecting the container in the annotation is likely to cause issues with the annotation parser cache. This is a very bad idea IMO

**Comment by stof** [comment]
@Ener-Getick because there is 2 ways to work with annotations:

- the old way using the constructor
- the newest way using public properties directly, for which the annotation parser can validate what are the supported properties (and also which one are required and the type of values)

File: Controller/Annotations/AbstractParam.php (hunk 1 above)
Comment (Ener-Getick) [inline Controller/Annotations/AbstractParam.php]: This behaviour was introduced by #1087 and I transformed it to introduce the ``ParamInterface``.

Should we keep this feature @lsmith77 ?

Comment (Ener-Getick) [inline Controller/Annotations/AbstractParam.php]: Firstly, I will remove this in this PR as this is a BC break ...
But we have to decide what to do for the master branch.

**Comment by Ener-Getick** [comment]
Ok so we can leave the annotations as is.

Should I deprecate or add BC layers for something else ?

**Comment by Ener-Getick** [comment]
``ParamFetcherInterface``, ``ParamReaderInterface``, ``ViewHandlerInterface`` and ``ConfigurableViewHandlerInterface`` are IMO not really useful and there are probably not used.
Should we not remove them to make the bundle clearer ?

**Comment by lsmith77** [comment]
@Ener-Getick I assume you went through the UPGRADING guide?

aside from this, here you can see all changes between master and 1.7:
https://github.com/FriendsOfSymfony/FOSRestBundle/compare/1.7...master

**Comment by Ener-Getick** [comment]
I didn't think to use the github comparator ^_^
Apparently it remains the context adapters. I will create a simple support (just the context classes not the symfony serializer support).

**Comment by xabbuh** [comment]
IIRC in Symfony we made some properties `protected` in annotations and added magic getters and setters to be able to trigger deprecation warnings.

By the way, what do you think about adding `@internal` annotations to a lot of classes? This would make it easier for us to break BC in the future (many classes actually should never be used directly by consumers of the bundle)?

File: Request/AbstractRequestBodyParamConverter.php [hunk 5]
```diff
@@ -151,21 +155,23 @@ protected function execute(Request $request, ParamConverter $configuration)
     }
 
     /**
-     * @return DeserializationContext
+     * @return DeserializationContext|Context
      */
     protected function getDeserializationContext()
```
Comment (Ener-Getick) [inline Request/AbstractRequestBodyParamConverter.php]: What about renaming these functions to ``getContext`` and ``setContext`` ?

File: Context/GroupableContextInterface.php [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+namespace FOS\RestBundle\Context;
+
+/**
+ * Context with groups.
+ *
+ * @author Ener-Getick <egetick@gmail.com>
+ */
+interface GroupableContextInterface extends ContextInterface
```
Comment (Ener-Getick) [inline Context/GroupableContextInterface.php]: I'm wondering if we should not remove the sub-context interfaces and focus on the most used features.

**Comment by Ener-Getick** [comment]
:+1: @xabbuh, keeping BC is sometimes really hard because of this classes...

File: Request/AbstractRequestBodyParamConverter.php (hunk 5 above)
Comment (xabbuh) [inline Request/AbstractRequestBodyParamConverter.php]: Not sure if we need them at all, when we introduce the serializer API with #1171. What do you think?

Comment (Ener-Getick) [inline Request/AbstractRequestBodyParamConverter.php]: These functions allow to define the (de)serialization context for a specific view. How would we do this with only the serializer api?

Comment (xabbuh) [inline Request/AbstractRequestBodyParamConverter.php]: Ah yes, you're right. Maybe we should backport the serializer API to 1.8, add our own layer for the context and rely on that then.

Comment (Ener-Getick) [inline Request/AbstractRequestBodyParamConverter.php]: Yeah probably better to import the 2.0 features to 1.8 when possible and then remove all the deprecations in 2.0.

**Comment by Ener-Getick** [comment]
I'll split this PR into smallest PRs, it begins to be too complicated to manage...

**Pull request closed without merging**